KEP-4671: Decoupled PodGroup and Workload API in alphav2 (#5893)
kubernetes/enhancements: keps/sig-scheduling/4671-gang-scheduling/README.md

Reviewer: ```suggestion
The kube-scheduler will be watching for `PodGroup` objects (using informers) and will use them to map pods
to and from their `PodGroup` objects.
```

Reviewer 2: Are we expecting to have status for the `Workload`? My assumtion is this API will be similar to `StorageClass` or any other config API.

In this KEP, kube-scheduler is modified to support gang scheduling[^1]. We focus on framework support and building blocks, not the ideal gang-scheduling algorithm - it can come as a follow-up. We start with simpler implementation of gang scheduling, kube-scheduler identifies pods that are in a group and waits until all pods reach the same stage of the scheduling/binding cycle before allowing any pods from the group to advance past that point.  If not all pods can reach that point before a timeout expires, then the scheduler stops trying to schedule that group, and all pods release all their resources.  This allows other workloads to try to allocate those resources.

New core types called `Workload` and `PodGroup` are introduced to tell the kube-scheduler that a group of pods 

> should be scheduled together and any policy options related to gang scheduling. Pods may have an object reference in 

Reviewer: ```suggestion
should be scheduled together and define policy options related to gang scheduling. Pods may have an object reference in 
```

their spec to the `PodGroup` they belong to. The `Workload` and `PodGroup` objects are intended to evolve[^2] via 
future KEPs to support additional kube-scheduler improvements, such as topology-aware scheduling.

In 1.36, we redesigned the API in order to clearly decouple Workload API from the runtime PodGroup API. The design 
for this can be found in the [^4] and the change itself is part of the separate `[KEP-5832](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/5832-decouple-podgroup-api/README.md).`

In the updated design, `Workload` represents a static template defining the scheduling hierarchy and policies. A 

> separate runtime object, `PodGroup`, represents the actual runtime instance of grouped pods. `Pods` reference their 

Author: PTAL at my version.

New core types called `Workload` and `PodGroup` are introduced to tell the kube-scheduler that a group of pods 
should be scheduled together and any policy options related to gang scheduling. Pods may have an object reference in 
their spec to the `PodGroup` they belong to. The `Workload` and `PodGroup` objects are intended to evolve[^2] via 
future KEPs to support additional kube-scheduler improvements, such as topology-aware scheduling.

In 1.36, we redesigned the API in order to clearly decouple Workload API from the runtime PodGroup API. The design 

> for this can be found in the [^4] and the change itself is part of the separate KEP-5832.

Author: Added link. It will not work though until the PR for KEP-5832 is merged.

Workloads that require gang scheduling often also need all members of the gang to be as topologically "close" to one another as possible, in order to perform adequately. Existing Pod affinity rules influence pod placement, but they do not consider the gang as a unit of scheduling and they do not cause the scheduler to efficiently try multiple mutually exclusive placement options for a set of pods. The design of the Workload object introduced in this KEP anticipates how Gang Scheduling support can evolve over subsequent KEPs into full Topology-aware scheduling support in kube-scheduler.

The `PodGroup` object will allow kube-scheduler to be aware that pods are part of a `Workload` with complex 
internal structure. Those workloads include builtins like `Job` and `StatefulSet`, and custom workloads, like 
`JobSet`, `LeaderWorkerSet`, `MPIJob` and `TrainJob`. All of these workload types are used for AI training and inference use cases.

> The 1.36 revision, as detailed in the KEP-5832 and [^4], addresses feedback regarding the ambiguity of the original 

Reviewer 3: This paragraph is a valid addition, but I don't think it belongs to Motivation - I think it rather belongs to the Proposal section.

## Proposal

The 1.36 revision, as detailed in the KEP-5832 and [^4], addresses feedback regarding the ambiguity of the original
"monolithic" Workload. By decoupling policy (Workload) from runtime grouping (PodGroup), we improve clarity,
scalability and lifecycle management. This approach also provides a more readable and intuitive structure for
complex workloads like JobSet and LeaderWorkerSet.

The `spec.workloadRef` field in the Pod API identifies the scheduling context. In the 1.36 implementation, a Pod is 
associated with a workload and a specific runtime instance. A sample pod with these new fields looks like this:

```yaml
apiVersion: v1
kind: Pod
spec:
  ...

>   workloadRef:

Reviewer 3: 1) Let's remove Pod->Workload reference (at least for now). This isn't strictly needed for correctness. The UX/debuggability isn't that important and I agree we can push for now.
I really like the idea of storing that in status but let's decouple it and maybe add that as future ideas.

The above would address the first concern.

2)  The struct is conceptually what we already have now and for be it boils down to naming. I like that (modulo naming, but naming we can hopefully figure out during code review).

  completionMode: Indexed
  template:
    spec:

>       schedulingGroup:

Author: Fair point. I agree that these group associations will likely impact the broader lifecycle (preemption, resource attachment, etc.), but I'm still a bit hesitant about the name workgroup. To me, it feels a bit too generic  – it doesn't immediately signal the purpose of the field to the user.

To me, 'scheduling' remains the primary functional anchor and the clearest entry point for users. It's the 'why' behind creating these groups in the first place.

I'm leaving it like this in KEP and I'm happy to continue the discussion during the API review. Implementation PRs will be submitted promptly to front-load the review cycle.

               "metadata.annotations['batch.kubernetes.io/job-completion-index']"
```

The `Workload` resource is a new core resource that provides scheduling policy definitions. It does not manage pod 
lifecycles or interfere with the pod creation logic of controllers like `Job`, `JobSet`, or `StatefulSet`. Instead, it 

> serves as a policy blueprint, describing the PodGroupTemplates with their corresponding scheduling policies

Author: You're right. I will update the text to make this clear: Workload is the template, PodGroup is self-contained (and completely enough for scheduler)

               "metadata.annotations['batch.kubernetes.io/job-completion-index']"
```

The `Workload` resource is a new core resource that provides scheduling policy definitions. It does not manage pod 
lifecycles or interfere with the pod creation logic of controllers like `Job`, `JobSet`, or `StatefulSet`. Instead, it 

> serves as a policy blueprint, describing to the scheduler which pod templates are expected and which policies (e.g., 

Reviewer 3: pod templates are out of scope for this KEP, so let's remove it.
Let's focus just on the scheduling policies.

### Associating Pod into PodGroups

> The WorkloadReference struct in PodSpec provides the linkage between the Pod and the Workload API objects. In 1.36,  

Author: Pending decision in https://github.com/kubernetes/enhancements/pull/5893/files#r2771034809. Will update accordingly once the decision is made.

	WorkloadRef *WorkloadReference
}

// WorkloadReference identifies the Workload object and PodGroup instance
// that a Pod belongs to. The scheduler uses this information to apply
// workload-aware scheduling semantics.
type WorkloadReference struct {
    // Name defines the name of the Workload object this Pod belongs to.
    // Workload must be in the same namespace as the Pod.
    // It must be a DNS subdomain.
    //
    // +optional
    Name string

    // PodGroup is tombstoned since the field in 1.36 was replaced with PodGroupName.
    // PodGroup string

    // PodGroupReplicaKey is tombstoned since the field in 1.36 was replaced with PodGroupName.
    // PodGroupReplicaKey string

    // PodGroupName is the name of the runtime PodGroup object that this Pod
    // belongs to. If it doesn't match any existing PodGroup,
    // the Pod will remain unschedulable until the PodGroup object is created
    // and observed by the kube-scheduler. It must be a DNS label.
    // +optional
    PodGroupName string `json:"podGroupName,omitempty"`
}
```

> Note: In 1.36, all fields in WorkloadReference are made optional, and the validation logic for the Name and 

Author: Rewritten. PTAL.

### Associating Pod into PodGroups

We propose introducing a `SchedulingGroup` field in `PodSpec` (replacing the previous `WorkloadReference`) to link
the `Pod` to its scheduling context.

```go
type PodSpec struct {
	...
	// SchedulingGroup provides a reference to the immediate scheduling runtime grouping object that this Pod 
	// belongs to. In the current implementation, this is always a PodGroup, but it may evolve in the future to support
	// other concepts like PodSubGroups.
	// This field is used by the scheduler to identify the PodGroup and apply the
	// correct group scheduling policies. The PodGroup object referenced
	// by this field may not exist at the time the Pod is created.
	// This field is immutable, but a PodGroup object with the same name
	// may be recreated with different policies. Doing this during pod scheduling
	// may result in the placement not conforming to the expected policies.
	//
	// +featureGate=GenericWorkload
	// +optional
	SchedulingGroup *PodSchedulingGroup
	

> 	// WorkloadRef is tombstoned since the field in 1.36 was replaced with SchedulingGroup.

Reviewer 3: nit: change the order to don't change the tags etc.

the `Pod` to its scheduling context.

```go
type PodSpec struct {
	...
	
    // WorkloadRef is tombstoned since the field in 1.36 was replaced with SchedulingGroup.
    // WorkloadRef *WorkloadReference
	
	// SchedulingGroup provides a reference to the immediate scheduling runtime grouping object that this Pod 
	// belongs to. In the current implementation, this is always a PodGroup, but it may evolve in the future to support
	// other concepts like PodSubGroups.
	// This field is used by the scheduler to identify the PodGroup and apply the
	// correct group scheduling policies. The PodGroup object referenced
	// by this field may not exist at the time the Pod is created.
	// This field is immutable, but a PodGroup object with the same name
	// may be recreated with different policies. Doing this during pod scheduling
	// may result in the placement not conforming to the expected policies.
	//
	// +featureGate=GenericWorkload
	// +optional
	SchedulingGroup *PodSchedulingGroup
}

// PodSchedulingGroup identifies the runtime scheduling group instance that a Pod belongs to. 
// The scheduler uses this information to apply workload-aware scheduling semantics.

> type PodSchedulingGroup struct {

Reviewer 4: I think this should be described as a one-of which only has one option.  What I mean is that it currently allows 3 states to exist:

This has clear meaning:
```
Pod:
  spec:
    schedulingGroup: null     // <--- not specified
```

This has clear meaning:
```
Pod:
  spec:
    schedulingGroup:
        podGroupName: "my-podgroup"
```

What does this mean?  Is it actually allowed?
```
Pod:
  spec:
    schedulingGroup: {}
```

We need to be super careful that we don't make "nothing" mean something, or else we can't ever evolve it (because we cannot tell the difference between "said nothing" and "said something I don't know how to decode").

So I think I would say that the inside of `PodSchedulingGroup` is a one-of.  Each field is optional but ONE OF THEM must be specified.  There just happens to only be one option.  For now.

}
```

Note: In 1.36, all fields in WorkloadReference are made optional, and the validation logic for the Name and 
PodGroupName fields presence is moved to the code to allow for greater flexibility if valid needs arise later.

At least for Alpha, we start with `WorkloadReference` to be immutable field in the Pod.
In further phases, we may decide to relax validation and allow for setting some of the fields later.
Moreover, the visibility into issues (debuggability) will depend on [#5501], but we don't
treat it as a blocker.

[#5501]: https://github.com/kubernetes/enhancements/pull/5501

The example below shows how this could look with the decoupled architecture for a simple job-like workload.

A Workload object defines the static PodGroup template:

```yaml
apiVersion: scheduling.k8s.io/v1alpha2
kind: Workload
metadata:
  name: jobset
spec:
  podGroupTemplates:
      - name: "job-1"
        schedulingPolicy:
        gang:
          minCount: 100
```

A standalone PodGroup object is created to define the scheduling policy and track a specific runtime instance:

```yaml
apiVersion: scheduling.k8s.io/v1alpha2
kind: PodGroup
metadata:
  name: job-instance-worker-0
spec:

>   workloadRef:

Reviewer 3: We changed the structure of referencing as decribed here:
https://github.com/kubernetes/enhancements/pull/5833#discussion_r2774073656

Can you please update to reflect it?

- for replicated gang, we can't use the full label selector, but rather support specifying only the
  label key, similar to `MatchLabelKeys` in pod affinity

> This decoupled model clearly separates the role of a Pod (runtime) from its blueprint (Workload). We decided on this

Reviewer: ```suggestion
Decoupling `Workload` from `PodGroup` (in 1.36) clearly separates the role of a Pod (runtime) from its blueprint (Workload). We decided on this
```

- for replicated gang, we can't use the full label selector, but rather support specifying only the
  label key, similar to `MatchLabelKeys` in pod affinity

> Decoupling `Workload` from `PodGroup` (in 1.36) clearly separates the role of a Pod (runtime grouping, status and 

Reviewer 3: s/Pod/PodGRoup/

	Name string
}

// PodGroupTemplate represents a blueprint for a set of pods with a common policy.
type PodGroupTemplate struct {

>   // Name is a unique identifier for the template within the Workload.

Reviewer 3: In general - let's try to minimize the diff here.
